amqp/src/main/scala/akka/stream/alpakka/amqp/model.scala

```diff
@@ -28,57 +31,110 @@ object ReadResult {
+    val mandatory: Boolean,
+    val passThrough: PT = NotUsed,
+    val properties: Option[BasicProperties] = None,
+    val routingKey: Option[String] = None
+) {
 
-  def withImmediate(value: Boolean): WriteMessage =
+  def withImmediate(value: Boolean): WriteMessage[PT] =
     if (value == immediate) this
     else copy(immediate = value)
 
-  def withMandatory(value: Boolean): WriteMessage =
+  def withMandatory(value: Boolean): WriteMessage[PT] =
     if (value == mandatory) this
     else copy(mandatory = value)
 
-  def withProperties(properties: BasicProperties): WriteMessage =
+  def withPassThrough[PT2](passThrough: PT2): WriteMessage[PT2] =
+    new WriteMessage[PT2](
+      bytes = bytes,
+      immediate = immediate,
+      mandatory = mandatory,
+      passThrough = passThrough,
+      properties = properties,
+      routingKey = routingKey
+    )
+
+  def withProperties(properties: BasicProperties): WriteMessage[PT] =
     copy(properties = Some(properties))
 
-  def withRoutingKey(routingKey: String): WriteMessage =
+  def withRoutingKey(routingKey: String): WriteMessage[PT] =
     copy(routingKey = Some(routingKey))
 
   private def copy(immediate: Boolean = immediate,
                    mandatory: Boolean = mandatory,
                    properties: Option[BasicProperties] = properties,
                    routingKey: Option[String] = routingKey) =
-    new WriteMessage(bytes, immediate, mandatory, properties, routingKey)
+    new WriteMessage(bytes, immediate, mandatory, passThrough, properties, routingKey)
 
   override def toString: String =
     "WriteMessage(" +
     s"bytes=$bytes, " +
     s"immediate=$immediate, " +
     s"mandatory=$mandatory, " +
+    s"passThrough=$passThrough, " +
     s"properties=$properties, " +
     s"routingKey=$routingKey" +
     ")"
 }
 
 object WriteMessage {
-  def apply(bytes: ByteString): WriteMessage =
+  def apply(bytes: ByteString): WriteMessage[NotUsed] =
     new WriteMessage(bytes, immediate = false, mandatory = false)
 
-  def apply(bytes: ByteString, immediate: Boolean, mandatory: Boolean): WriteMessage =
+  def apply(bytes: ByteString, immediate: Boolean, mandatory: Boolean): WriteMessage[NotUsed] =
     new WriteMessage(bytes, immediate, mandatory)
 
   /**
    * Java API
    */
-  def create(bytes: ByteString): WriteMessage = WriteMessage(bytes)
+  def create(bytes: ByteString): WriteMessage[NotUsed] = WriteMessage(bytes)
 
   /**
    * Java API
    */
-  def create(bytes: ByteString, immediate: Boolean, mandatory: Boolean): WriteMessage =
+  def create(bytes: ByteString, immediate: Boolean, mandatory: Boolean): WriteMessage[NotUsed] =
     WriteMessage(bytes, immediate, mandatory)
 }
+
+final class WriteResult[T] private (val confirmed: Boolean, val passThrough: T) {
+  def rejected: Boolean = !confirmed
+
+  override def toString: String =
+    s"WriteResult(confirmed=$confirmed, passThrough=$passThrough)"
+
+  override def equals(other: Any): Boolean = other match {
```

Comment:
`PartialFunction.cond`

Reply:
It doesn't make it much simpler, and every single equals method in whole alpakka seems to be implemented using `match`, so I'll leave that as is.